examples/generate.go

```diff
@@ -0,0 +1,234 @@
+}
+
+type exampleOutput struct {
+	Alg        jose.SignatureAlgorithm
+	WIT        string
+	WITClaims  witClaims
+	Key        ed25519.PublicKey
+	PrivateKey ed25519.PrivateKey
+}
+
+func generateWorkloadIdentityToken(claims witClaims, inputOptions []inputOption, outputOptions []outputOption) (exampleOutput, error) {
+	input := exampleInput{
+		rand: rand.Reader,
+		alg:  jose.EdDSA,
+	}
+	for _, opt := range inputOptions {
+		if err := opt(&input); err != nil {
+			return exampleOutput{}, fmt.Errorf("failed to apply option: %w", err)
+		}
+	}
+
+	pubKey, privKey, err := ed25519.GenerateKey(input.rand)
+	if err != nil {
+		panic("failed to generate wl key")
+	}
+
+	claims.Cnf = cnf{
+		JWK: jose.JSONWebKey{
+			Key:       pubKey,
+			Algorithm: string(input.alg),
+		},
+	}
+	cb := jwt.Signed(input.signer).Claims(claims)
+	enc, err := cb.Serialize()
+	if err != nil {
+		return exampleOutput{}, fmt.Errorf("failed to sign WIT: %w", err)
+	}
+
+	result := exampleOutput{
+		Alg:        input.alg,
+		WIT:        enc,
+		WITClaims:  claims,
+		Key:        pubKey,
+		PrivateKey: privKey,
+	}
+	for _, opt := range outputOptions {
+		if err := opt(&result); err != nil {
+			return exampleOutput{}, fmt.Errorf("failed to apply option: %w", err)
+		}
+	}
+	return result, nil
+}
+
+type inputOption func(input *exampleInput) error
+
+func withRand(rand io.Reader) inputOption {
+	return func(input *exampleInput) error {
+		input.rand = rand
+		return nil
+	}
+}
+
+func withSigner(signer jose.Signer) inputOption {
+	return func(input *exampleInput) error {
+		input.signer = signer
+		return nil
+	}
+}
+
+type outputOption func(result *exampleOutput) error
+
+func withHttpMessageSigRequest(req *http.Request, outSignatureInput *string, outSignature *string) outputOption {
+	return func(result *exampleOutput) error {
+		req.Header["Workload-Identity-Token"] = []string{result.WIT}
+
+		sigConfig := httpsign.NewSignConfig()
+		sigConfig.SetTag("wimse-workload-to-workload")
+		sigConfig.SetNonce("abcd1111")
+		sigConfig.SignAlg(false)
+		sigConfig.SetExpires(time.Now().Add(time.Minute).Unix())
+		signer, err := httpsign.NewEd25519Signer(result.PrivateKey,
+			sigConfig,
+			httpsign.Headers("@method", "@request-target", "workload-identity-token"))
+		if err != nil {
+			return fmt.Errorf("failed to create request signer: %w", err)
+		}
+		signatureInput, signature, err := httpsign.SignRequest("wimse", *signer, req)
+		if err != nil {
+			return fmt.Errorf("failed to sign request: %w", err)
+		}
+
+		*outSignatureInput = signatureInput
+		*outSignature = signature
+		return nil
+	}
+}
+
+func withHttpMessageSigResponse(req *http.Request, resp *http.Response, outSignatureInput *string, outSignature *string) outputOption {
+	return func(result *exampleOutput) error {
+		resp.Header["Workload-Identity-Token"] = []string{result.WIT}
+
+		sigConfig := httpsign.NewSignConfig()
+		sigConfig.SetTag("wimse-workload-to-workload")
+		sigConfig.SetNonce("abcd2222")
```

Comment:
hex encoding of 128 bits of deterministic random data is needed for a jti but "abcd2222" is good enough?